RHDEVDOCS-3491 - Document support for Workload Resource Mapping (#46315)
openshift/openshift-docs: modules/sbo-binding-workloads-that-are-not-compliant-with-PodSpec.adoc

[id="sbo-binding-workloads-that-are-not-compliant-with-PodSpec_{context}"]
= Binding secondary workloads that are not compliant with PodSpec

A typical scenario in service binding involves configuring the backing service, the workload (Deployment), and {servicebinding-title}. Consider a scenario that involves a secondary workload (which can also be an application Operator) that is not compliant with PodSpec and is between the primary workload (Deployment) and {servicebinding-title}.

For such secondary workload resources, the location of the container path is arbitrary. For service binding, if the secondary workload in a CR is not compliant with the PodSpec, you must specify the location of the container path. Doing so projects the binding data into the container path specified in the secondary workload of the `ServiceBinding` custom resource (CR), for example, when you do not want the binding data inside a pod.

> {servicebinding-title} provides an option to configure the value of where the container path or secret path is and bind these paths at a custom location.

Reviewer: - Does this mean you can configure **where** the path information is (the location of the path information), or that you can configure the **path** itself? 
- Does this mean you can bind these paths *at* a custom location (control where the binding happens), or that you can bind these paths **to** a custom location?

Can we update the wording to answer these questions and rephrase the information in terms of the user's needs? 

What about:
```suggestion
In {servicebinding-title}, you can configure the container path or secret path, and bind these paths to a custom location.
```
OR 
```suggestion
In {servicebinding-title}, you can configure the location of the container path or secret path, and bind these paths at a custom location.
```

Author: Preferred version:
In {servicebinding-title}, you can configure the path of where containers or secrets reside within a workload and bind these paths at a custom location.

Author: Fixed

[id="configuring-custom-location-of-container-path_{context}"]

> == Configuring the custom location of the container path

Reviewer: Is the word "custom" here significant? Or, would the meaning be the same without it? For example:
```suggestion
== Configuring the location of the container path
```

> This option is available only for the `binding.operators.coreos.com` API group when the binding data is projected as environment variables.

Reviewer: - The word `only` is playing games with my understanding of this sentence. Can we remove `only` and keep the same meaning?
- Can we restate the name of the option, custom location, instead of saying "This option"?
- Can we rephrase "is projected" using the active voice?

How about:
```suggestion
This custom location is available for the `binding.operators.coreos.com` API group when {servicebinding-title} projects the binding data as environment variables.
```

Reviewer: Can you rephrase this abstract in terms of the user's goals and actions? You can typically produce content that is much more user-centered by applying the following template to the first sentence in each section:
`To <goal>, you can <action> by using <feature>.`

Author: Fixed

<1> Unique array of containers with values generated by the {servicebinding-title}. These values are based on the backing service CR.
<2> Name of the `Secret` resource generated by the {servicebinding-title}.

[id="configuring-custom-location-of-secret-path_{context}"]
== Configuring the custom location of the secret path

> This option is available only for the `binding.operators.coreos.com` API group when the binding data is projected as environment variables.

Reviewer: Similar questions here.

Author: Fixed

    secret: binding-request-72ddc0c540ab3a290e138726940591debf14c581 <1>
...
----

> <1> Unique name of the `Secret` resource generated by the {servicebinding-title}.

Reviewer: ```suggestion
<1> The unique name of the `Secret` resource that {servicebinding-title} generates.
```
- Use articles and the present tense.
- Avoid the passive voice.

Reviewer: Does the qualifier, `generated by the {servicebinding-title}` help the user identify the `Secret` resource? If not, consider dropping it.

Author: Fixed

[id="workload-resource-mapping_{context}"]
== Workload resource mapping

[NOTE]
====

> * Workload resource mapping is available for the secondary workloads of the ServiceBinding custom resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.

Reviewer: ```suggestion
* Workload resource mapping is available for the secondary workloads of the `ServiceBinding` custom resource (CR) for both the API groups: `binding.operators.coreos.com` and `servicebinding.io`.
```

Author: Fixed

> * You must namespace `ClusterWorkloadResourceMapping` resources only under the `servicebinding.io` API group. However, the `ClusterWorkloadResourceMapping` resources interact with `ServiceBinding` resources under both the `binding.operators.coreos.com` and `servicebinding.io` API groups.

Author: * You must define `ClusterWorkloadResourceMapping` resources only under the `servicebinding.io` API group. However, the `ClusterWorkloadResourceMapping` resources interact with `ServiceBinding` resources under both the `binding.operators.coreos.com` and `servicebinding.io` API groups.

> * You must namespace `ClusterWorkloadResourceMapping` resources only under the `servicebinding.io` API group. However, the `ClusterWorkloadResourceMapping` resources interact with `ServiceBinding` resources under both the `binding.operators.coreos.com` and `servicebinding.io` API groups.

Author: Fixed

====

> The specification also provides a way to define exactly where binding data needs to be projected. Use this method when you are not able to configure custom path locations correctly by previous methods. You can define `ClusterWorkloadResourceMapping` resources only in the `servicebinding.io` API group, which specify where binding data is to be projected for a given workload kind.

Reviewer: The following suggestion embodies the following 
- Focus on the user perspective.
- Simplify "you are not able to" to "you cannot"
- Start with the precondition; "If you cannot..."
- Tell the user where to perform the task instead of stating that the task can "only" be performed in a specific location. 

How about:
```suggestion
If you cannot configure custom path locations by using the previous methods, you can define exactly where binding data needs to be projected. Specify where to project the binding data for a given workload kind by defining the `ClusterWorkloadResourceMapping` resources in the `servicebinding.io` API group.
```
- Can we make "previous methods" more specific?

Author: If you cannot configure custom path locations by using the configuration method for container path, you can define exactly where binding data needs to be projected. Specify where to project the binding data for a given workload kind by defining the `ClusterWorkloadResourceMapping` resources in the `servicebinding.io` API group.

.Example: Schema of the `ClusterWorkloadResourceMapping` resource
[source,yaml]
----
apiVersion: servicebinding.io/v1alpha3
kind: ClusterWorkloadResourceMapping
metadata:
  name: # string <1>
  ...
spec:
  versions: # required, at least one entry must be defined
  - version: # string <2>
    annotations: # string (Fixed JSONPath), optional <3>
    containers: # optional
    - path: # string (JSONPath) <4>
      name: # string (Fixed JSONPath), optional <5>
      env: # string (Fixed JSONPath), optional <6>
      volumeMounts: # string (Fixed JSONPath), optional <7>

>     volumes: # string (Fixed JSONPath), optional <8>

Reviewer: In examples, please use example values or `<replaceable strings in brackets>`. Avoid comments like this.

Reviewer: Replaceable values should follow this format: `<ip_address>`

Reviewer: I don't see another schema in the openshift-docs repo that uses comments this way. Please use replacement values and callouts in the typical fashion instead.

Reviewer: We avoid repeating because over time, one set of information might get updated while the other doesn't. This produces confusing or conflicting information and maintenance headaches.

Reviewer 2: Discussed this with @Srivaralakshmi.  We've replaced this with a more minimal example.

<4> Identifier of the `.containers` and `.initContainers` fields in a pod, specified with a JSONPath.  If no entries under the `containers` field are defined, the {servicebinding-title} defaults to two paths: `.spec.template.spec.containers[\*]` and `.spec.template.spec.initContainers[\*]`, with all other fields set as their default.  However, if you specify an entry, then you must define the `.path` field.
<5> Optional: Identifier of the `.name` field in a container, specified with a fixed JSONPath.  The default value is `.name`.
<6> Optional: Identifier of the `.env` field in a container, specified with a fixed JSONPath.  The default value is `.env`.
<7> Optional: Identifier of the `.volumeMounts` field in a container, specified with a fixed JSONPath.  The default value is `.volumeMounts`.
<8> Optional: Identifier of the `.volumes` field in a pod, specified with a fixed JSONPath.  The default value is `.spec.template.spec.volumes`.

[IMPORTANT]
====
* In this context, a fixed JSONPath is a subset of the JSONPath grammar that accepts only the following operations:

** Field lookup: `.spec.template`
** Array indexing: `.spec['template']`
+

> All other operations are forbidden.

Reviewer: Can we rephrase what's "forbidden" in terms of what we support or don't support?

Author: Fixed

> All other operations are forbidden.

Author: ```suggestion
All other operations are not accepted.
```

Reviewer: If the other operations generate a "forbidden" error, you can say forbidden. Otherwise, avoid it as patronizing.

* Most of these fields are optional. When they are not specified, the {servicebinding-title} assumes defaults compatible with `PodSpec` resources.

> * The {servicebinding-title} requires that each of these fields is structurally equivalent to the corresponding field in a pod deployment.  For example, the contents of the `.env` field in a workload resource must be able to accept the same structure of data that the `.env` field in a Pod resource would.  Otherwise, projecting binding data into such a workload might result in unexpected behavior from the {servicebinding-title}.

Author: Fixed

====

Using the previous schema, you can define an example mapping for the `CronJob.batch/v1` resources.

.Example: Mapping for `CronJob.batch/v1` resources
[source,yaml]
----
apiVersion: servicebinding.io/v1alpha3
kind: ClusterWorkloadResourceMapping
metadata:
 name: cronjobs.batch
spec:
  versions:
  - version: "v1"
    annotations: .spec.jobTemplate.spec.template.metadata.annotations
    containers:
    - path: .spec.jobTemplate.spec.template.spec.containers[*]
    - path: .spec.jobTemplate.spec.template.spec.initContainers[*]
    volumes: .spec.jobTemplate.spec.template.spec.volumes
----

> [NOTE]

Reviewer: How does this information relate to the preceding task? Can you rephrase it so that the user can relate it to the previous task or what they might need to do next?  If not, does it belong in a separate concept module?

> [NOTE]

Author: .Behavior specific to the `binding.operators.coreos.com` API group

You can expect the following behaviors when `ClusterWorkloadResourceMapping` resources interact with `ServiceBinding` resources under the `binding.operators.coreos.com` API group:

* If a `ServiceBinding` resource with the `bindAsFiles: false` flag value is created together with one of these mappings, then environment variables are projected into the `.envFrom` field underneath each `path` field specified in the corresponding `ClusterWorkloadResourceMapping` resource.
* As a cluster administrator, you can specify both a `ClusterWorkloadResourceMapping` resource and the `.spec.application.bindingPath.containersPath` field in a `ServiceBinding.bindings.coreos.com` resource for binding purposes.
+
The {servicebinding-title} attempts to project binding data into the locations specified in both a `ClusterWorkloadResourceMapping` resource and the `.spec.application.bindingPath.containersPath` field.  This behavior is equivalent to adding a container entry to the corresponding `ClusterWorkloadResourceMapping` resource with the `path: $containersPath` attribute, with all other values taking their default value.
